### ISSUE TITLE
How to declare that an exception is thrown

### ISSUE THREAD

**Issue by HoneyryderChuck** [ruby/rbs#421]
I didn't find in the documentation how to declare if an exception can be thrown. Is this something that will be supported?

**Comment by soutaro** [comment]
We don't have a plan to support it for now. (Future version may have it, but not in Ruby 3.0.)

**Comment by fxn** [comment]
Let me share one possible notation.

My main motivation for writing method signatures is documentation and help understanding of the source code when you read it. So, not coming from the type-checking camp, but there is a lot of overlap.

In particular, I annotate exceptions, because as a user of the code, you need to know what a method may raise.

Have recently adopted the `#:` notation in Zeitwerk (love it is so succint) and have extended the current syntax:

```ruby
#: (String | Pathname, namespace: Module) -> void ! Zeitwerk::Error
def push_dir(path, namespace: Object)
  ...
end
```

I have chosen `!` because it is one character, looks like a separator (vertical-ish line), its associated meaning in Ruby resonates, and it is used in the `$!` variable, which holds exception objects in `rescue` clauses.

**Comment by fxn** [comment]
One possible downside is that `!` and `|` are not super different, in practice it should be seen if returned union types may get confused or need extra carefulness when reading the signature. So there's that, but at the same time `!` seems like the right choice. What do you think?

**Comment by pointlessone** [comment]
Make it a sigil on the type: ` -> void | !Zeitwerk::Error`

Exception is an early return, after all. We just need to make the distinction between a returned type and a risen type.

Since we're here, do we want to have a different notation for `throw`? It's also a type of an early return but also different to an exception.

```
() -> void | !Zeitwerk::Error | ~Integer
```

I kinda like `^` for denoting throws (up the stack) but I guess we can throw a proc so that would be confusing.

Just a though.

**Comment by ParadoxV5** [comment]
Interesting. A method (or Proc) can have multiple raise and throw types. For example,
```ruby
def example(x) = case x
when 1..10
  throw x, -x
when 11..20
  throw x.to_f, x.to_s
when 21..30
  raise ArgumentError
when 31..40
  raise MyError
end
```
* What are we looking at here?
  ```rbs
  def example: (Integer) -> ! ArgumentError ! MyError ~ Integer => Integer ~ Float => String
  ```
* or could it be? (with #1645)
  ```rbs
  def example: (Integer) -> (ArgumentError, MyError, Integer => Integer, Float => String)
  ```

**Comment by fxn** [comment]
@ParadoxV5 exactly, it was implicit in the proposal that a list of possible exceptions would be expressed as a union.

**Comment by pointlessone** [comment]
Can you throw multiple things? According to the docs `throw` take a symbol (tag), and an optional value. And when you catch the specified tag you can get that value. So you type that value as any other return type. Probably a union of all possible things the value can be.

Though, now that I refreshed on the throw, we have to specify the tag and it has to be a literal symbol. I guess we can reuse that syntax but it's probably better to use a union here as well. I'm not sure about the tag with a value. A whole record is probably an overkill. Maybe some reduced syntax.

Same goes for exceptions. You can't raise multiple exceptions.

So:

```ruby
def example: (Integer) -> void ! ArgumentError | MyError ~ tag1: Integer | tag2: Float | :bare_tag3
```

**Comment by ParadoxV5** [comment]
The tag need not to be a Symbol, and the optional value one can consider as defaulting to `nil`.

**Comment by fxn** [comment]
I have been exchanging impressions with @paracycle and @Morriar today.

One possible alternative would be to list exceptions after the regular signature:

```ruby
#: (String | Pathname, namespace: Module) -> void
#! ArgumentError | Zeitwerk::Error
def push_dir(path, namespace: Object)
  ...
end
```

That seems easier to read, and using a union type allows also for one compact form that has it all. If a method raises something, you have it in one line.

It has to be seen how does that work with alternative signatures using `#|`, guess you'd need to have a corresponding `#!` for each, since on paper you could be also raising different sets of exceptions in each one.

**Comment by pointlessone** [comment]
1. A syntax for .rbs files is still needed
2. I personally am not a fan of even more special comment syntaxes. I'd rather we had a common comment annotations among all Ruby tools. I'm for `# @annotation` syntax (e.g. `# @rbs () -> void`) but that's like just my opinion.

**Comment by fxn** [comment]
@pointlessone I have used `@raise` in my annotations for years, but I was using my own `@sig` tag.

I am playing with this change trying to think about an alternative precisely because `@raise` together with `#:` looks odd to me. They are documenting basically the formal method interface using completely different notation styles.

Since I don't use type-checkers or RBS generation, I can bend the rules a little bit.

**Comment by fxn** [comment]
@pointlessone oh, let me add that I'd be OK with `@rbs` instead of `#:` too, and then tags are a uniform theme. Well, unless that implies one line per param in YARD-style, which I am personally not a fan of (prefer terse signatures and vertical space economy).

On the other hand, I believe tags have a tendency to get out of hand, because it is so easy to define a new one: `@return`, `@see_also`, `@attr`, `@deprecated`, `@api`, `@author`, .... All of a sudden, you have stacks of tags taking more space than the code itself. By choosing something more constrained, you kind of limit that. 

So, there is a tension. At the same time, a signature that does not include raised exceptions is incomplete, in my view.

I was not there, but I believe the current trend towards `#:` was initiated at the last RubyKaigi, in which there was a votation. Is that right @soutaro?

**Comment by fxn** [comment]
You are right. Was also thinking that method overloads could have their own sets of exceptions, on paper. That may not work as nicely if exceptions go in their own line

```ruby
#: (Integer) -> void
#! StandardError | CustomError
#| (Float) -> void
#! StandardError | CustomError2
```

Not sure if `#|` is a thing, I believe I saw it, but don't see where. In any case, we might still be getting a readability problem there due to how similar `!` and `|` are, you need to pay attention, and I think that is not desirable.
